File: 백준/Silver/2805. 나무 자르기/나무 자르기.java
import java.util.Arrays;
import java.util.Scanner;

public class Main {
    public long count(int capacity, int[] arr){
        long treeLength=0;
        for (int x: arr){
            if(x-capacity>0){
                treeLength += (x-capacity);
            }
//            System.out.println(treeLength);
        }
        return treeLength;
    }
    public int solution(int n, int m, int[] arr){
        int answer=0;
        int lt = 1;
        int rt = Arrays.stream(arr).max().getAsInt();

        int mid;
        long treeLength;
        while(lt<=rt){
            mid = (lt+rt)/2;
            treeLength = count(mid, arr);
//            System.out.println("capacity:"+mid+"treeLength:"+treeLength);
            if(treeLength>=m){
                answer = mid;
                lt = mid+1;
            }else{
                rt = mid-1;
            }
        }
        return answer;
    }
    public static void main(String[] ars){
        Main T = new Main();
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();
        int[] arr = new int[n];
        for(int i=0;i<n;i++){
            arr[i] = sc.nextInt();
        }
        System.out.print(T.solution(n,m,arr));

    }
}

File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
import sys

n, m = map(int, sys.stdin.readline().strip().split())
h = list(map(int, sys.stdin.readline().strip().split()))

answer = 0
start = 0
end = max(h)


while start <= end:
    mid = (start + end) // 2
    s = 0

    for i in h:
        if i - mid > 0:
            s += (i-mid)

    if s >= m:
        answer = mid
        start = mid + 1
    else:
        end = mid - 1
print(answer)
